server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -264,20 +750,64 @@ public List<HealthIndicatorImpact> getImpacts() {
         }
 
+        /**
+         * Summarizes the user actions that are needed to solve unassigned primary and replica shards.
+         * @param includeDetails true if user actions should be generated, false if they should be omitted.
+         * @return A summary of user actions. Alternatively, an empty list if none were found or includeDetails is false.
+         */
+        public List<UserAction> getUserActions(boolean includeDetails) {
+            if (includeDetails) {
+                Map<UserAction.Definition, Set<String>> actionsToAffectedIndices = new HashMap<>(primaries.userActions);
+                replicas.userActions.forEach((actionDefinition, indicesWithReplicasUnassigned) -> {
+                    Set<String> indicesWithPrimariesUnassigned = actionsToAffectedIndices.get(actionDefinition);
+                    if (indicesWithPrimariesUnassigned == null) {
+                        actionsToAffectedIndices.put(actionDefinition, indicesWithReplicasUnassigned);
+                    } else {
+                        indicesWithPrimariesUnassigned.addAll(indicesWithReplicasUnassigned);
+                    }
+                });
+                if (actionsToAffectedIndices.isEmpty()) {
+                    return Collections.emptyList();
+                } else {
+                    return actionsToAffectedIndices.entrySet()
+                        .stream()
+                        .map(
+                            e -> new UserAction(
+                                e.getKey(),
+                                e.getValue().stream().sorted(byPriorityThenByName(clusterMetadata)).collect(Collectors.toList())
+                            )
+                        )
+                        .collect(Collectors.toList());
+                }
+            } else {
+                return Collections.emptyList();
+            }
+        }
     }
 
     private static String getTruncatedIndicesString(Set<String> indices, Metadata clusterMetadata) {
         final int maxIndices = 10;
-        final int unknownIndexPriority = -1;
-        // We want to show indices with a numerically higher index.priority first (since lower priority ones might get truncated):
-        Comparator<String> priorityThenNameComparator = Comparator.comparingInt((String indexName) -> {
-            IndexMetadata indexMetadata = clusterMetadata.index(indexName);
-            return indexMetadata == null ? unknownIndexPriority : indexMetadata.priority();
-        }).reversed().thenComparing(Comparator.naturalOrder());
-        String truncatedIndicesString = indices.stream().sorted(priorityThenNameComparator).limit(maxIndices).collect(joining(", "));
+        String truncatedIndicesString = indices.stream()
+            .sorted(byPriorityThenByName(clusterMetadata))
+            .limit(maxIndices)
+            .collect(joining(", "));
         if (maxIndices < indices.size()) {
             truncatedIndicesString = truncatedIndicesString + ", ...";
```

Comment:
I am halfway tempted to suggest we replace this with "…" but I think that would be a bit too far :D